File: lecture_007(Strings)/l_011_printsubsequence.java
import java.util.Scanner;
public class l_011_printsubsequence {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        String s = scn.nextLine();
        printSubsequence(s);
    }
    public static void printSubsequence(String s) {
        for(int i=0; i< (1<<s.length()) ; i++) {
            for(int j=s.length()-1; j>=0; j--) {
                int mask = 1<<j;
                if( (i&mask) != 0 ) {     //jth bit is on
                    System.out.print(s.charAt(s.length()-1-j));   
                }
                else {
                    System.out.print("_");
                }
            }
            System.out.println();
        }
    }
}

File: lecture_007(Strings)/l_011_printsubsequence.cpp
#include<iostream>
#include<string>
using namespace std;

void printsubsequence(string s) {
    for(int i=0; i< (1<<s.size()); i++) {
        for(int j=s.size()-1; j>=0; j--) {
            int mask = 1<<j;
            if( (i&mask) != 0 ) {
                cout<<s[s.size()-1-j];
            }
            else {
                cout<<"-";
            }
        }
        cout<<endl;
    }
}
int main() {
    string s;
    cin>>s;
    printsubsequence(s);

    return 0;
}